django_irods/views.py

```diff
@@ -0,0 +1,477 @@
+logger = logging.getLogger(__name__)
+
+
+def download(request, path, rest_call=False, use_async=True, use_reverse_proxy=True,
+             *args, **kwargs):
+    """ perform a download request, either asynchronously or synchronously
+
+    :param request: the request object.
+    :param path: the path of the thing to be downloaded.
+    :param rest_call: True if calling from REST API
+    :param use_async: True means to utilize asynchronous creation of objects to download.
+    :param use_reverse_proxy: True means to utilize NGINX reverse proxy for streaming.
+
+    The following variables are computed:
+
+    * `path` is the public path of the thing to be downloaded.
+    * `irods_path` is the location of `path` in irods.
+    * `output_path` is the output path to be reported in the response object.
+    * `irods_output_path` is the location of `output_path` in irods
+
+    and there are seven cases:
+
+    1. path points to a single file that is not a single-file-aggregation
+       object in a composite resource.
+    2. path points to a single file that is a single-file-aggregation
+       object in a composite resource.
+    3. path points to a metadata object that may need updating.
+    4. path points to a bag that needs to be updated and then returned.
+    5. path points to a folder that needs to be zipped.
+    6. path points to a previously zipped file that was zipped asynchronously.
+    7. path points to a .zip file that -- when the .zip is removed -- is a file that can be zipped
+
+    In cases 1, 3, 4, and 6 the output path is the input path.
+    In cases 2, 5, and 7, the output path is created as a result of
+    the operation of zipping the object and its metadata into a new zipfile.
+    """
+    logger.debug("request path is {}".format(path))
+
+    split_path_strs = path.split('/')
+    while split_path_strs[-1] == '':
+        split_path_strs.pop()
+    path = u'/'.join(split_path_strs)  # no trailing slash
+
+    # initialize case variables
+    is_bag_download = False
+    is_zip_download = False
+    is_zip_request = False
+    is_sf_agg_file = False
+
+    if split_path_strs[0] == 'bags':
+        is_bag_download = True
+        # format is bags/{rid}.zip
+        # logger.debug("fetching rid from bags field 1 = {}".format(split_path_strs[1]))
+        res_id = os.path.splitext(split_path_strs[1])[0]
+    elif split_path_strs[0] == 'zips':
+        is_zip_download = True
+        # zips prefix means that we are following up on an asynchronous download request
+        # format is zips/{date}/{zip-uuid}/{public-path}.zip where {public-path} contains the rid
+        # logger.debug("fetching rid from zips field 3 = {}".format(split_path_strs[3]))
+        res_id = split_path_strs[3]
+    else:  # regular download request
+        # logger.debug("fetching rid from public path field 0 = {}".format(split_path_strs[0]))
+        res_id = split_path_strs[0]
+
+    logger.debug("resource id is {}".format(res_id))
```

Comment:
Remove this debug log?

Reply:
Later, after we are sure this works properly. See comment above.